hummingbot/connector/exchange/bitmax/bitmax_api_order_book_data_source.py

```diff
@@ -0,0 +1,260 @@
+            return [convert_from_exchange_trading_pair(item["symbol"]) for item in data["data"]]
+
+    @staticmethod
+    async def get_order_book_data(trading_pair: str) -> Dict[str, any]:
+        """
+        Get whole orderbook
+        """
+        async with aiohttp.ClientSession() as client:
+            resp = await client.get(f"{REST_URL}/depth?symbol={convert_to_exchange_trading_pair(trading_pair)}")
+            if resp.status != 200:
+                raise IOError(
+                    f"Error fetching OrderBook for {trading_pair} at {EXCHANGE_NAME}. "
+                    f"HTTP status is {resp.status}."
+                )
+
+            data: List[Dict[str, Any]] = await safe_gather(resp.json())
+            item = data[0]
+            if item.get("code") != 0:
+                raise IOError(
+                    f"Error fetching OrderBook for {trading_pair} at {EXCHANGE_NAME}. "
+                    f"Error is {item.message}."
+                )
+
+            return item["data"]
+
+    async def get_new_order_book(self, trading_pair: str) -> OrderBook:
+        snapshot: Dict[str, Any] = await self.get_order_book_data(trading_pair)
+        snapshot_timestamp: float = snapshot.get("data").get("ts")
+        snapshot_msg: OrderBookMessage = BitmaxOrderBook.snapshot_message_from_exchange(
+            snapshot.get("data"),
+            snapshot_timestamp,
+            metadata={"trading_pair": trading_pair}
+        )
+        order_book = self.order_book_create_function()
+        active_order_tracker: BitmaxActiveOrderTracker = BitmaxActiveOrderTracker()
+        bids, asks = active_order_tracker.convert_snapshot_message_to_order_book_row(snapshot_msg)
+        order_book.apply_snapshot(bids, asks, snapshot_msg.update_id)
+        return order_book
+
+    async def listen_for_trades(self, ev_loop: asyncio.BaseEventLoop, output: asyncio.Queue):
+        while True:
+            try:
+                trading_pairs = ",".join(list(
+                    map(lambda trading_pair: convert_to_exchange_trading_pair(trading_pair), self._trading_pairs)
+                ))
+                payload = {
+                    "op": "sub",
+                    "ch": f"trades:{trading_pairs}"
+                }
+
+                async with websockets.connect(WS_URL) as ws:
+                    ws: websockets.WebSocketClientProtocol = ws
+                    await ws.send(ujson.dumps(payload))
+
+                    async for raw_msg in self._inner_messages(ws):
+                        try:
+                            msg = ujson.loads(raw_msg)
+                            if (msg is None or msg.get("m") != "trades"):
+                                continue
+
+                            trading_pair: str = convert_from_exchange_trading_pair(msg.get("symbol"))
+
+                            for trade in msg.get("data"):
+                                trade_timestamp: int = ms_timestamp_to_s(trade.get("ts"))
```

Comment:
Might I ask why there is a need to convert the timestamps to seconds? It seems all the responses from the REST API as well as the WebSocket API are all in ms.

Reply:
For some reason, I always thought HB operated is seconds! Fixed.